hub3/ead/dao.go

```diff
@@ -0,0 +1,622 @@
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *DaoClient) PublishFindingAid(cfg *DaoConfig) error {
+	fa, err := cfg.FindingAid(c)
+	if err != nil {
+		return err
+	}
+
+	cfg.updateRevisionKey()
+
+	for _, file := range fa.Files {
+		fg, err := cfg.fragmentGraph(file)
+		if err != nil {
+			return err
+		}
+		fg.Meta.SourceID = cfg.RevisionKey
+
+		m, err := fg.IndexMessage()
+		if err != nil {
+			return err
+		}
+
+		if c.bi != nil {
+			c.bi.Publish(context.Background(), m)
+		}
+	}
+
+	fg, err := cfg.findingAidFragmentGraph(&fa)
+	if err != nil {
+		return err
+	}
+	fg.Meta.SourceID = cfg.RevisionKey
+
+	m, err := fg.IndexMessage()
+	if err != nil {
+		return err
+	}
+
+	if c.bi != nil {
+		c.bi.Publish(context.Background(), m)
+	}
+
+	cfg.ObjectCount = int(fa.GetFileCount())
+	cfg.MimeTypes = getMimeTypes(&fa)
+
+	return c.dropOrphans(cfg)
+}
+
+func (c *DaoClient) StoreMets(cfg *DaoConfig) error {
+	log.Debug().
+		Str("link", cfg.Link).
+		Str("orgID", cfg.OrgID).
+		Str("datasetID", cfg.ArchiveID).
+		Str("InventoryID", cfg.InventoryID).
+		Msg("storing remote mets file")
+
+	resp, err := c.client.Get(cfg.Link)
+	if err != nil {
+		metsRetrieveErr := fmt.Errorf("unable to retrieve METS %s client error: %s", cfg.Link, err)
+		logMETSError(cfg.ArchiveID, cfg.InventoryID, metsRetrieveErr.Error())
+
+		return metsRetrieveErr
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		metsStatusErr := fmt.Errorf("unable to retrieve METS %s HTTP status error: %d", cfg.Link, resp.StatusCode)
+		logMETSError(cfg.ArchiveID, cfg.InventoryID, metsStatusErr.Error())
+
+		return metsStatusErr
+	}
+
+	defer resp.Body.Close()
+
+	mets, err := metsParse(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	mets.CmetsHdr.AttrCREATEDATE = ""
+	mets.CmetsHdr.AttrLASTMODDATE = ""
+
+	var buf bytes.Buffer
+	enc := xml.NewEncoder(&buf)
+	enc.Indent("", "\t")
+
+	if err := enc.Encode(mets); err != nil {
+		return err
+	}
+
+	if _, err := os.Stat(cfg.getMetsFilePath()); os.IsNotExist(err) {
+		if mkDirErr := os.MkdirAll(cfg.getDirPath(), os.ModePerm); mkDirErr != nil {
+			return mkDirErr
+		}
+	}
+
+	return ioutil.WriteFile(
+		cfg.getMetsFilePath(),
+		buf.Bytes(),
+		os.ModePerm,
+	)
+}
+
+func (c *DaoClient) DefaultDaoFn(cfg *DaoConfig) error {
+	if !strings.Contains(cfg.Link, "/gaf/api/mets/v1/") {
+		return fmt.Errorf("invalid daolink to GAF: ' %s '", cfg.Link)
+	}
+
+	return c.PublishFindingAid(cfg)
+}
+
+func validateMetsRequest(r *http.Request) (string, string, error) {
+	spec := chi.URLParam(r, "spec")
+	err := ValidateSpec(spec)
+	if err != nil {
+		return "", "", err
+	}
+
+	uuid := chi.URLParam(r, "UUID")
+	if uuid == "" {
+		return "", "", fmt.Errorf("UUID cannot be empty")
+	}
+
+	return spec, uuid, nil
+}
+
+// Index indexes the stored METS files identified by their UUID
+func (c *DaoClient) Index(w http.ResponseWriter, r *http.Request) {
+	spec, uuid, err := validateMetsRequest(r)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
```

Comment:
Is it ok to not return after the error as is done in similar situations such as HandleDelete?

Reply:
added the return